### PR TITLE
Switch to cmake for build

### PR DESCRIPTION
**Pull request by mthrok**
Supersedes #1068

### CODE REVIEW

File: third_party/transducer/CMakeLists.txt [hunk 1]
```diff
@@ -1,27 +1,3 @@
...
-        SET(CMAKE_SKIP_RPATH TRUE)
-    ENDIF()
-
-ELSE()
-    # always skip for linux
-    SET(CMAKE_SKIP_RPATH TRUE)
-ENDIF()
-
-ADD_LIBRARY(warprnnt STATIC submodule/src/rnnt_entrypoint.cpp)
+add_library(warprnnt STATIC submodule/src/rnnt_entrypoint.cpp)
```
Comment (cpuhrsch) [inline third_party/transducer/CMakeLists.txt:1]: We should probably also hide warprnnt symbols

File: torchaudio/csrc/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    libtorchaudio
+    )
+
+  set(TORCHAUDIO_LIBRARY -Wl,--no-as-needed libtorchaudio -Wl,--as-needed CACHE INTERNAL "")
+endif()
+
+################################################################################
+# _torchaudio.so
+################################################################################
+if (BUILD_PYTHON_EXTENSION)
```
Comment (cpuhrsch) [inline torchaudio/csrc/CMakeLists.txt:52]: suggestion: I'd go with BUILD_TORCHAUDIO_PYTHON_EXTENSION just because this is quite generic

**Review by cpuhrsch — APPROVED** [review]

File: third_party/transducer/CMakeLists.txt (hunk 1 above)
Comment (mthrok) [reply third_party/transducer/CMakeLists.txt:1]: I misspoke during the review. `-fvisibility=hidden` is defined on `third_party` level so it is propagated to transducer module too.

```
$ objdump -t build/temp.linux-x86_64-3.8/third_party/transducer/libwarprnnt.a | grep hidden
0000000000000000 g     F .text	0000000000000006 .hidden get_warprnnt_version
0000000000000010 g     F .text	0000000000000038 .hidden rnntGetStatusString
0000000000000050 g     F .text	000000000000005e .hidden get_workspace_size
0000000000000000  w    F .text._ZN7CpuRNNTIfE14compute_alphasEPKfiiPf	00000000000002b8 .hidden _ZN7CpuRNNTIfE14compute_alphasEPKfiiPf
0000000000000000  w    F .text._ZN7CpuRNNTIdE14compute_alphasEPKdiiPd	000000000000029d .hidden _ZN7CpuRNNTIdE14compute_alphasEPKdiiPd
0000000000000000  w    F .text._ZN7CpuRNNTIfE22compute_betas_and_gradEPfPKfiiS1_S1_PKif	00000000000008fd .hidden _ZN7CpuRNNTIfE22compute_betas_and_gradEPfPKfiiS1_S1_PKif
0000000000000000  w    F .text._ZN7CpuRNNTIfE13cost_and_gradEPKfPfS3_PKiS5_S5_	000000000000051f .hidden _ZN7CpuRNNTIfE13cost_and_gradEPKfPfS3_PKiS5_S5_
00000000000000b0 g     F .text	0000000000000610 .hidden compute_rnnt_loss
0000000000000000  w    F .text._ZN7CpuRNNTIdE22compute_betas_and_gradEPdPKdiiS1_S1_PKid	00000000000008b7 .hidden _ZN7CpuRNNTIdE22compute_betas_and_gradEPdPKdiiS1_S1_PKid
0000000000000000  w    F .text._ZN7CpuRNNTIdE13cost_and_gradEPKdPdS3_PKiS5_S5_	00000000000004c4 .hidden _ZN7CpuRNNTIdE13cost_and_gradEPKdPdS3_PKiS5_S5_
00000000000006c0 g     F .text	00000000000005f0 .hidden compute_rnnt_loss_fp64
0000000000000000         *UND*	0000000000000000 .hidden __dso_handle
```

Comment (mthrok) [reply third_party/transducer/CMakeLists.txt:1]: The thing is that, to build `libsox`, we use `ExternalProject_Add` which does not reflect the regular CMake configuration like `CMAKE_CXX_FLAGS` so it has to be set again there.

**Comment by mthrok** [comment]
THANK YOU!

**Pull request merged**